pkg/ring/replication_set_tracker_test.go

```diff
@@ -0,0 +1,266 @@
+			instances:           []IngesterDesc{instance1, instance2, instance3, instance4, instance5, instance6},
+			maxUnavailableZones: 1,
+			run: func(t *testing.T, tracker *zoneAwareResultTracker) {
+				// Track failing instances.
+				for _, instance := range []IngesterDesc{instance1, instance2} {
+					tracker.done(&instance, errors.New("test"))
+					assert.False(t, tracker.succeeded())
+					assert.False(t, tracker.failed())
+				}
+
+				// Track successful instances.
+				for _, instance := range []IngesterDesc{instance3, instance4, instance5} {
+					tracker.done(&instance, nil)
+					assert.False(t, tracker.succeeded())
+					assert.False(t, tracker.failed())
+				}
+
+				tracker.done(&instance6, nil)
+				assert.True(t, tracker.succeeded())
+				assert.False(t, tracker.failed())
+			},
+		},
+		"should succeed as soon as the response has been successfully received from 'all zones - 1' on max unavailable zones = 1": {
+			instances:           []IngesterDesc{instance1, instance2, instance3, instance4, instance5, instance6},
+			maxUnavailableZones: 1,
+			run: func(t *testing.T, tracker *zoneAwareResultTracker) {
+				// Track successful instances.
+				for _, instance := range []IngesterDesc{instance1, instance2, instance3} {
+					tracker.done(&instance, nil)
+					assert.False(t, tracker.succeeded())
+					assert.False(t, tracker.failed())
+				}
+
+				tracker.done(&instance4, nil)
+				assert.True(t, tracker.succeeded())
+				assert.False(t, tracker.failed())
+			},
+		},
+		"should succeed on failing instances within 2 zones on max unavailable zones = 2": {
+			instances:           []IngesterDesc{instance1, instance2, instance3, instance4, instance5, instance6},
+			maxUnavailableZones: 2,
+			run: func(t *testing.T, tracker *zoneAwareResultTracker) {
+				// Track failing instances.
+				for _, instance := range []IngesterDesc{instance1, instance2, instance3, instance4} {
+					tracker.done(&instance, errors.New("test"))
+					assert.False(t, tracker.succeeded())
+					assert.False(t, tracker.failed())
+				}
+
+				// Track successful instances.
+				tracker.done(&instance5, nil)
+				assert.False(t, tracker.succeeded())
+				assert.False(t, tracker.failed())
+
+				tracker.done(&instance6, nil)
+				assert.True(t, tracker.succeeded())
+				assert.False(t, tracker.failed())
+			},
+		},
+		"should succeed as soon as the response has been successfully received from 'all zones - 2' on max unavailable zones = 2": {
+			instances:           []IngesterDesc{instance1, instance2, instance3, instance4, instance5, instance6},
+			maxUnavailableZones: 2,
+			run: func(t *testing.T, tracker *zoneAwareResultTracker) {
+				// Zone-a
+				tracker.done(&instance1, nil)
+				assert.False(t, tracker.succeeded())
+				assert.False(t, tracker.failed())
+
+				// Zone-b
+				tracker.done(&instance3, nil)
+				assert.False(t, tracker.succeeded())
+				assert.False(t, tracker.failed())
+
+				// Zone-a
+				tracker.done(&instance1, nil)
```

Comment:
Shouldn't this be instance2?

Reply:
Ops, yes! Fixed. Test was passing anyway because trackers don't keep track of "which" instance has done, only the count.